add server directIO

cclhsu opened the pull request:
#### Proposed Changes ####

- add server directIO.
- fix error when send/receive file size less than block size (4096) using directIO. 
- Add/Update unit test.

#### Types of Changes ####

enhancement

#### Verification ####

- Add/Update unit test.

#### Linked Issues ####

Refs: # https://github.com/longhorn/longhorn/issues/1943

#### Further Comments ####

None

Signed-off-by: Clark Hsu <clark.hsu@suse.com>

shuo-wu reviewed (changes requested):
~~When the sender starts to send a file, I think it should tell the server if it can use direct IO:~~
~~https://github.com/longhorn/sparse-tools/blob/c1877f4a574f57a461f7a69709defd20bc52718a/sparse/client.go#L73~~

~~https://github.com/longhorn/sparse-tools/blob/c1877f4a574f57a461f7a69709defd20bc52718a/sparse/rest/handlers.go#L51-L70~~

shuo-wu commented on sparse/client.go:
```diff
@@ -144,6 +147,7 @@ func (client *syncClient) sendHTTPRequest(method string, action string, interval
 	}
 
 	req.Header.Add("Accept", "application/json")
+	req.Header.Add("DirectIO", strconv.FormatBool(directIO))
```
Can we put this parameter in `req.URL.Query()`?

cclhsu replied:
changed

shuo-wu commented on sparse/client.go:
```diff
@@ -184,12 +188,12 @@ func (client *syncClient) openServer() error {
 
-func (client *syncClient) closeServer() {
-	client.sendHTTPRequest("POST", "close", Interval{0, 0}, nil)
+func (client *syncClient) closeServer(directIO bool) {
```
It seems that the server won't check this new parameter `directIO` and actually it is no need to check it. In other words, we don't need to modify the function signature.

Same for the below `syncClient.getServerChecksum` and `syncClient.writeData`.

cclhsu replied:
changed.

shuo-wu commented on sparse/client.go:
```diff
@@ -127,7 +130,7 @@ func (client *syncClient) syncFileContent(file FileIoProcessor, fileSize int64)
 }
 
-func (client *syncClient) sendHTTPRequest(method string, action string, interval Interval, data []byte) (*http.Response, error) {
+func (client *syncClient) sendHTTPRequest(method string, action string, interval Interval, data []byte, directIO bool) (*http.Response, error) {
```
The new parameter `directIO` is required by the caller `client.openServer(directIO)` only. Asking other callers to set this option doesn't make sense. Hence, can we modify the input like the following?
```
sendHTTPRequest(method string, action string, querys map[string]string, data []byte)
```

cclhsu replied:
changed

cclhsu commented:
@shuo-wu I have updated the code according the review and integation test in https://github.com/longhorn/longhorn-engine/pull/560 has passed.

shuo-wu commented on sparse/client.go:
```diff
@@ -127,7 +130,7 @@ func (client *syncClient) syncFileContent(file FileIoProcessor, fileSize int64)
 }
 
-func (client *syncClient) sendHTTPRequest(method string, action string, interval Interval, data []byte) (*http.Response, error) {
+func (client *syncClient) sendHTTPRequest(method string, action string, interval Interval, queries map[string]string, data []byte) (*http.Response, error) {
```
I mean `func (client *syncClient) sendHTTPRequest(method string, action string, queries map[string]string, data []byte) (*http.Response, error)`

All callers need to put `interval.Begin` and `interval.End` in `queries` before calling this function. Then This function can blindly put all entries of `queries` in `req.URL.Query()` without `directIO, err := queries["directIO"]`

cclhsu replied:
updated

shuo-wu commented on sparse/rest/handlers.go:
```diff
@@ -22,6 +22,26 @@ type SyncFileStub struct{}
 func (f *SyncFileStub) UpdateSyncFileProgress(size int64) {}
 
+func (server *SyncServer) getQueryDirectIO(request *http.Request) (bool, error) {
+	var directIO bool
+	var err error
+
+	queryParams := request.URL.Query()
+	directIOStr := queryParams.Get("directIO") // only one value for key directIO
+	if directIOStr == "" {
+		err = fmt.Errorf("queryParams directIO does not exist")
+		return directIO, err
+	}
+
+	directIO, err = strconv.ParseBool(directIOStr)
+	if err != nil {
+		err = fmt.Errorf("strconv.ParseBool(directIOStr) error: %s", err)
+		return directIO, err
+	}
+
+	return directIO, err
```
I would prefer this kind of style:
```suggest
	queryParams := request.URL.Query()
	directIOStr := queryParams.Get("directIO")
	if directIOStr == "" {
		return false, fmt.Errorf("directIO does not exist in the queries: %+v", queryParams)
	}

	directIO, err := strconv.ParseBool(directIOStr)
	if err != nil {
		err = fmt.Errorf("strconv.ParseBool(directIOStr) error: %s", err)
		return false, errors.Warpf(err, "failed to parse directIO %v", directIOStr)
	}
	return directIO, nil
```

Returning `return directIO, err` in the last line is a little bit confusing.

cclhsu replied:
updated

shuo-wu commented on sparse/client.go:
```diff
@@ -146,23 +149,28 @@ func (client *syncClient) sendHTTPRequest(method string, action string, interval
 	timeStart := time.Now()
 	timeStop := timeStart.Add(time.Duration(client.timeout) * time.Second)
+	queries := make(map[string]string)
+	queries["begin"] = strconv.FormatInt(0, 10)
+	queries["end"] = strconv.FormatInt(client.fileSize, 10)
```
Don't need to add these 2 entries. Since the server won't invoke `getQueryInterval` after receiving `open` calls.

Same for `func (client *syncClient) closeServer() {`

cclhsu replied:
updated

shuo-wu replied:
OH, sorry. This suggestion should be applied to `closeServer` only. And you already found that.

shuo-wu commented on sparse/rest/handlers.go:
```diff
@@ -22,6 +22,23 @@ type SyncFileStub struct{}
 func (f *SyncFileStub) UpdateSyncFileProgress(size int64) {}
 
+func (server *SyncServer) getQueryDirectIO(request *http.Request) (bool, error) {
+	var directIO bool
+	var err error
```
redundant if you modify line 35 to the following: 
```
directIO, err := strconv.ParseBool(directIOStr)
```

cclhsu replied:
updated

Commit:
add command line argument directio rev 2 (#1943)

add command line argument directio to resolve ticket '[BUG] sparse-tools: ssync should explicitly asking for directIO or not, rather than determine is using filesize' (#1943)

Signed-off-by: Clark Hsu <clark.hsu@suse.com>

shuo-wu reviewed (approved):
LGTM
